atlasdb-client/src/test/java/com/palantir/atlasdb/transaction/impl/TransactionRetryStrategyTest.java

```diff
@@ -0,0 +1,135 @@
+import static org.mockito.ArgumentMatchers.anyLong;
+import static org.mockito.Mockito.when;
+
+import java.util.List;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.function.IntPredicate;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import com.github.rholder.retry.BlockStrategy;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import com.palantir.atlasdb.transaction.api.TransactionFailedNonRetriableException;
+import com.palantir.atlasdb.transaction.api.TransactionFailedRetriableException;
+import com.palantir.atlasdb.transaction.impl.TransactionRetryStrategy.Retryable;
+
+@RunWith(MockitoJUnitRunner.class)
+public class TransactionRetryStrategyTest {
+    private final TestBlockStrategy blockStrategy = new TestBlockStrategy();
+    @Mock private ThreadLocalRandom random;
+    @Mock private Retryable<String, Exception> task;
+    @Mock private IntPredicate shouldStopRetrying;
+
+    private TransactionRetryStrategy legacy;
+    private TransactionRetryStrategy exponential;
+
+    private static final class TestBlockStrategy implements BlockStrategy {
+        private long totalBlockedTime;
+
+        @Override
+        public void block(long sleepTime) {
+            totalBlockedTime += sleepTime;
+        }
+    }
+
+    @Before
+    public void before() {
+        when(random.nextLong(anyLong())).thenAnswer(inv -> (long) inv.getArgument(0) - 1);
+        when(shouldStopRetrying.test(anyInt())).thenAnswer(inv -> ((int) inv.getArgument(0)) > 1);
+        legacy = TransactionRetryStrategy.createLegacy(blockStrategy);
+        exponential = TransactionRetryStrategy.createExponential(blockStrategy, random);
+
+    }
+
+    private String run() throws Exception {
+        return exponential.runWithRetry(shouldStopRetrying, task);
+    }
+
+    @Test
+    public void retriesIfFailsWithRetriableException() throws Exception {
+        when(task.run()).thenThrow(new TransactionFailedRetriableException("")).thenReturn("success");
+        assertThat(run()).isEqualTo("success");
+    }
+
+    @Test
+    public void stopsIfShouldStopRetrying() throws Exception {
+        TransactionFailedRetriableException second = new TransactionFailedRetriableException("second");
+        when(task.run()).thenThrow(new TransactionFailedRetriableException("first")).thenThrow(second);
+        assertThatExceptionOfType(TransactionFailedRetriableException.class)
+                .isThrownBy(this::run)
+                .withMessage("Failing after 2 tries.")
+                .withCause(second);
+    }
+
+    @Test
+    public void doesNotRetryOnNonRetriableTransactionFailedException() throws Exception {
+        TransactionFailedNonRetriableException failure = new TransactionFailedNonRetriableException("");
+        when(task.run()).thenThrow(failure);
```

Comment:
can we add a successful response after this? shows that we don't bother retrying after the fact

Reply:
think you'll find that Mockito disallows that?

Reply:
it makes sense to add a verify